cmd/syncthing/relays.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (C) 2015 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/url"
+	"time"
+
+	syncthingprotocol "github.com/syncthing/protocol"
+	"github.com/syncthing/relaysrv/client"
+	"github.com/syncthing/relaysrv/protocol"
+	"github.com/syncthing/syncthing/internal/config"
+
+	"github.com/thejerf/suture"
+)
+
+func newRelayService(cfg *config.Wrapper, tlsCfg *tls.Config, conns chan<- connection) *relaySvc {
+	svc := &relaySvc{
+		Supervisor: suture.New("relaySvc", suture.Spec{
+			Log: func(log string) {
+				if debugNet {
+					l.Infoln(log)
+				}
+			},
+			FailureBackoff:   5 * time.Minute,
+			FailureDecay:     float64((10 * time.Minute) / time.Second),
+			FailureThreshold: 5,
+		}),
+		cfg:    cfg,
+		tlsCfg: tlsCfg,
+		conns:  conns,
+
+		invitations: make(chan protocol.SessionInvitation),
+	}
+
+	cfg.Subscribe(svc)
+
+	return svc
+}
+
+type relaySvc struct {
+	*suture.Supervisor
+	cfg    *config.Wrapper
+	tlsCfg *tls.Config
+	conns  chan<- connection
+
+	stop    chan struct{}
+	stopped chan struct{}
+
+	tokens      map[string]suture.ServiceToken
+	invitations chan protocol.SessionInvitation
+	running     bool
+}
+
+func (s *relaySvc) Serve() {
+	if debugNet {
+		l.Debugln("Relay service starting")
+	}
+	s.Supervisor.ServeBackground()
```

Comment:
Yeah this whole starting and stopping thing is hard. 
Checking for nil channels etc just seems like a hack. 

Though there is a teardown phase, so I guess I cannot just wrap it in a serviceFunc, plus it implements the config comitter so can't be a stateless service func.